File: cses/leetcode/go/34.go
package main

func searchRange(nums []int, target int) []int {
    ans := make([]int,2)
    n := len(nums)
    //finding the lower-bound of occurence
    leftValue := n
    for l,r := 0,n-1; l <= r; {
        mid := (l+r)/2
        if  (nums[mid]==target){
            leftValue = min(leftValue,mid)
            l = mid-1
        }
        else if(nums[mid]<target){
            r = mid+1
        }
    }
    return ans  
}

File: leetcode/cpp/34.cpp

#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    int binSearch(vector<int> nums, int target, bool findLowerBound)
    {
        int left = 0, right = nums.size() - 1;
        int index = -1;
        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            if (target > nums[mid])
            {
                left = mid + 1;
            }
            else if (target < nums[mid])
            {
                right = mid - 1;
            }
            else
            {
                if (findLowerBound)
                {
                    index = mid;
                    right = mid - 1;
                }
                else
                {
                    index = mid;
                    left = mid + 1;
                }
            }
        }
        return index;
    }
    vector<int> searchRange(vector<int> &nums, int target)
    {
        int lower = binSearch(nums, target, true);
        int upper = binSearch(nums, target, false);
        return {lower, upper};
    }
};
int main()
{
    vector<int> nums = {5, 7, 7, 8, 8, 10};
    int target = 8;
    Solution A = Solution();
    A.searchRange(nums, target);
}